modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/KeysTrackingConfigurationVisitor.java

```diff
@@ -195,32 +202,53 @@ private void endVisit(int previousKeyLength) {
         currentKey.setLength(previousKeyLength);
 
         currentPath.remove(currentPath.size() - 1);
-        legacyNames.remove(legacyNames.size() - 1);
+
+        String[] legacyNames = currentLegacyNames.remove(currentLegacyNames.size() - 1);
+        currentLegacyNamesCount -= legacyNames.length;
     }
 
-    /** Calls consumer for all legacy paths. */
-    protected void processLegacyPaths(ArrayList<String> path, BiConsumer<String, String> legacyKeyNewKeyConsumer) {
-        if (path.size() == currentPath().size() - 1) {
-            for (String legacyName : legacyNames.get(currentPath().size() - 1)) {
-                String legacyKey = join(appendKey(path, legacyName));
-                String newKey = currentKey();
+    /** Calls consumer for all variations of legacy paths, i.e. to remove them from the storage. */
+    protected void processLegacyPaths(Consumer<String> legacyKeyConsumer) {
+        // Current path doesn't contain any legacy names.
+        if (currentLegacyNamesCount == 0) {
+            return;
+        }
 
-                legacyKeyNewKeyConsumer.accept(legacyKey, newKey);
-            }
+        // Contains all combinations of current and old names.
+        var pathVariations = new ArrayList<List<String>>(currentPath.size());
```

Comment:
I don't really get why you need to create a separate list if all the information is already accessible in `currentPath` and `currentLegacyNames`.

Also, legacy names inside of named lists won't be handled correctly, as far as I see. This part is missing in deprecated configurations as well, this means that we need to create new Jira and add a huge TODO somewhere here. This is dangerous. Could you please do it?